zerver/tests/test_messages.py

```diff
@@ -1066,6 +1067,21 @@ def test_message_mentions(self) -> None:
         message = most_recent_message(user_profile)
         assert(UserMessage.objects.get(user_profile=user_profile, message=message).flags.mentioned.is_set)
 
+    def test_user_group_mentions(self) -> None:
+        iago = self.example_user('iago')
+        hamlet = self.example_user('hamlet')
+        zoe = self.example_user("ZOE")
+        create_user_group('user_group_a', [iago], get_realm('zulip'))
+        create_user_group('user_group_b', [iago, hamlet, zoe], get_realm('zulip'))
+        self.subscribe(iago, 'Denmark')
+        self.subscribe(hamlet, 'Denmark')
+        self.send_stream_message(self.example_user('cordelia'), 'Denmark',
+                                 content='Hi, @*user_group_a* @*user_group_b* @**Zoe**')
+        message = most_recent_message(iago)
+        assert(UserMessage.objects.get(user_profile=iago, message=message).flags.mentioned.is_set)
+        assert(UserMessage.objects.get(user_profile=hamlet, message=message).flags.mentioned.is_set)
+        assert(UserMessage.objects.get(user_profile=zoe, message=message).flags.mentioned.is_set)
+
```

Comment:
This test doesn't really test anything new; it passes even without the functional changes made in this commit. So I think that we can get rid of this.

Reply:
In fact, @hashirsarwar could you improve the test coverage for this commit? It doesn't seem like you're actually testing your change anywhere (but feel free to correct me if I'm wrong). The main point of this commit is that we will now look at why the user is being mentioned and not just if they're being mentioned (which we were already doing). So we need new tests to make sure that the reason the user is being mentioned is correct for different possible message.

Reply:
@Hypro999 I was intended to test my changes here. But yes, you are right, just looking for the `mentioned` flag isn't enough. We should probably look for [the `users` dict](https://github.com/zulip/zulip/pull/15011/files#diff-96ea148dcd22c60ffdb097f113d5e4aaR1488) to check which group is being mentioned for which user. Does that make sense?

Reply:
Um, the link you provided doesn't point to the exact thing that you're mentioning, could you fix it?

Reply:
@Hypro999 sorry, I just realized that. Updated the comment.

Reply:
You're definitely on the right track. This list of dictionaries is passed to `send_event` and then later `queue_json_publish` and so on. So directly observing this list can't be done without a workaround. Luckily the python standard library's testing framework gives us the ability to [
mock](https://docs.python.org/3/library/unittest.mock.html) function calls. So I think that mocking `send_event` might be the right approach to performing the test.